src/monitor/group_state_machine.c

```diff
@@ -570,6 +783,506 @@ ProceedGroupStateForPrimaryNode(AutoFailoverNode *primaryNode)
+	 * been demoted or is draining now. Most probably it's dead.
+	 *
+	 * Before we enter the selection process, we must have collected the last
+	 * received LSN from ALL the standby nodes that are considered as a
+	 * candidate (thanks to the FSM transition secondary -> report_lsn), and
+	 * now we need to select one of the failover candidates.
+	 *
+	 * We might fail to have reportedLSNCount == candidateCount in the
+	 * following cases:
+	 *
+	 * - the activeNode is the first to report its LSN, we didn't hear from the
+	 *   other nodes yet; in that case we want to wait until we hear from them,
+	 *   it is expected to happen within the next 5 seconds.
+	 *
+	 * - a standby that was counted as a candidate is not reporting; most
+	 *   probably it's not healthy anymore (or has been put to maintenance
+	 *   during the failover process), and the next call to node_active()
+	 *   should be able to account for that new global state and make progress:
+	 *   the now faulty standby will NOT be counted as candidate anymore.
+	 */
+	if (candidateList.waitingForReportedLSN > 0)
+	{
+		char message[BUFSIZE];
+
+		LogAndNotifyMessage(
+			message, BUFSIZE,
+			"Failover still in progress after %d nodes reported their LSN "
+			"and we are waiting for %d nodes to report, "
+			"activeNode is %d (%s:%d) and reported state \"%s\"",
+			candidateList.candidateCount, candidateList.waitingForReportedLSN,
+			activeNode->nodeId, activeNode->nodeHost, activeNode->nodePort,
+			ReplicationStateGetName(activeNode->reportedState));
+
+		return false;
+	}
+
+	/*
+	 * All the currently healthy standby nodes have reported their LSN, proceed
+	 * to find a node to failover to. Well, if we have at list one candidate.
+	 */
+	if (candidateList.candidateCount > 0)
+	{
+		/* find the standby node that has the most advanced LSN */
+		AutoFailoverNode *mostAdvancedNode =
+			FindMostAdvancedStandby(standbyNodesGroupList);
+
+		/* select a node to failover to */
+		AutoFailoverNode *selectedNode =
+			SelectFailoverCandidateNode(candidateList.candidateNodesGroupList,
+										mostAdvancedNode,
+										primaryNode);
+
+		/* we might not have a selected candidate for failover yet */
+		if (selectedNode == NULL)
+		{
+			/*
+			 * Publish more information about the process in the monitor event
+			 * table. This is a quite complex mechanism here, and it should be
+			 * made as easy as possible to analyze and debug.
+			 */
+			char message[BUFSIZE];
+
+			LogAndNotifyMessage(
+				message, BUFSIZE,
+				"Failover still in progress after all %d candidate nodes "
+				"reported their LSN and we failed to select one of them; "
+				"activeNode is %d (%s:%d) and reported state \"%s\"",
+				candidateList.candidateCount,
+				activeNode->nodeId, activeNode->nodeHost, activeNode->nodePort,
+				ReplicationStateGetName(activeNode->reportedState));
+
+			return false;
+		}
+
+		/*
+		 * Ok so we now may start the failover process, we have selected a
```

Comment:
To me, moving quite a bit logic to `BuildCandidateList ` made this code a lot more easier to follow it,  appreciate for that.

One further simplification could be to move the following logic into a function as well:

```C
Promote(S)ElectedNode(selectedNode, mostAdvancedNode, primaryNode)
{
... same code here..


}
```

 I'd be really happy if you agree on this. With that, the code would be trivially easy to follow after glance at the new FSM pdf you generated.

Reply:
Done now.